urllib3/connection.py

```diff
@@ -299,7 +302,15 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            try:
+                match_hostname(cert, self.assert_hostname or hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                e.cert = cert
+                raise e
```

Comment:
Yeah, this is an ongoing downside of the wrapped exceptions in requests: trying to *actually use* something that is in one of the lower exceptions is hard. However, my suspicion is that this mostly actually happens in some kind of semi-interactive setting, where people can drill down.

For the automated case, the log is probably sufficient. Having both covers all the bases. =)